nexus/reconfigurator/planning/src/planner.rs

```diff
@@ -328,70 +329,121 @@ impl<'a> Planner<'a> {
             }
         }
 
-        self.ensure_correct_number_of_nexus_zones(&sleds_waiting_for_ntp_zone)?;
-
-        Ok(())
+        self.do_plan_add_discretionary_zones(&sleds_waiting_for_ntp_zone)
     }
 
-    fn ensure_correct_number_of_nexus_zones(
+    fn do_plan_add_discretionary_zones(
         &mut self,
         sleds_waiting_for_ntp_zone: &BTreeSet<SledUuid>,
     ) -> Result<(), Error> {
-        // Count the number of Nexus zones on all in-service sleds. This will
-        // include sleds that are in service but not eligible for new services,
-        // but will not include sleds that have been expunged or decommissioned.
-        let mut num_total_nexus = 0;
-        for sled_id in self.input.all_sled_ids(SledFilter::InService) {
-            let num_nexus = self.blueprint.sled_num_nexus_zones(sled_id);
-            num_total_nexus += num_nexus;
-        }
+        // We usually don't need to construct an `OmicronZonePlacement` to add
+        // discretionary zones, so defer its creation until it's needed.
+        let mut zone_placement = None;
+
+        for zone_kind in [
+            DiscretionaryOmicronZone::Nexus,
+            DiscretionaryOmicronZone::CockroachDb,
+        ] {
+            // Count the number of `kind` zones on all in-service sleds. This
```

Comment:
What do you think about moving this block to another function? Might make it easier to test/call.

Reply:
I originally had it that way, but the `zone_placement` option that's reused across loop iterations was a little awkward. I've done some other rework since then; let me try it again...

Reply:
I tried extracting the full body of this loop into a separate function, and didn't like two things:

* The aforementioned `&mut Option<OmicronZonePlacement>` it has to take
* The function ended with an unconditional call to `add_discretionary_zones`, which felt like I was artificially splitting things up into two, which doesn't leave them independently testable

I took a second shot in `81b78ee64`: I kept the `Option<OmicronZonePlacement>` (and its construction, if needed) in the body of this loop, but moved the calculation of how many zones are needed into a new method. I think I like this better than either what was here before or moving the whole body of the loop into a method - what do you think?